segmentstore/server/src/main/java/io/pravega/segmentstore/server/ContainerEventProcessor.java

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.segmentstore.server;
+
+import io.pravega.common.util.BufferView;
+import lombok.Data;
+import lombok.NonNull;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+/**
+ * The {@link ContainerEventProcessor} is a sub-component in a Segment Container that aims at providing a durable,
+ * FIFO-like queue abstraction over an internal, system-critical Segment. The {@link ContainerEventProcessor} can manage
+ * one or more {@link EventProcessor}s, which are the ones that append events to the queue and handle events read. The
+ * also {@link ContainerEventProcessor} reports metrics for all the registered {@link EventProcessor}s.
+ */
+public interface ContainerEventProcessor extends AutoCloseable {
+
+    /**
+     * Instantiates a new {@link EventProcessor}. If the internal Segment exists, the {@link EventProcessor} will re-use
+     * it. If not, a new internal Segment will be created. Multiple calls to this method for the same name should result
+     * in returning the same {@link EventProcessor} object.
+     *
+     * @param name     Name of the {@link EventProcessor} object.
+     * @param handler  Function that will be invoked when one or more events have been read from the internal Segment.
+     * @param config   {@link EventProcessorConfig} for this {@link EventProcessor}.
+     * @return A {@link CompletableFuture} that, when completed, returns a new {@link EventProcessor} object associated
+     * to its own internal Segment.
+     */
+    CompletableFuture<EventProcessor> forConsumer(@NonNull String name,
+                                                  @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                                  @NonNull EventProcessorConfig config);
+
+    /**
+     * Instantiates a new {@link EventProcessor} that only enables to add new data to the internal Segment. This may be
+     * useful when a user needs to store events, but it is not clear how the best approach to process them will be. This
+     * append-only {@link EventProcessor} can be replaced in the future by a regular one with a handler function that
+     * will consume all the events in the internal Segment. If internal Segment exists, the {@link EventProcessor} will
+     * re-use it. If not, a new internal Segment will be  created. Multiple calls to this method for the same name
+     * should result in returning the same {@link EventProcessor} object.
+     *
+     * @param name     Name of the {@link EventProcessor} object.
+     * @return A {@link CompletableFuture} that, when completed, returns a new {@link EventProcessor} object associated
+     * to its own internal Segment.
+     */
+    CompletableFuture<EventProcessor> forDurableQueue(@NonNull String name);
```

Comment:
It will return the same `EventProcessor` object to all the callers, but I have added a test called `testConcurrentForConsumerCall()` for further testing.